File: packages/engine/test/parseLine.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { parseLine } from "../src/sources/claude-code/parseLine";

const fixturePath = join(__dirname, "fixtures", "sample-session.jsonl");
const fixtureLines = readFileSync(fixturePath, "utf-8")
  .split("\n")
  .filter((line) => line.trim().length > 0);

describe("parseLine", () => {
  it("returns null for non-JSON lines", () => {
    expect(parseLine("not json")).toBeNull();
  });

  it("returns null for blank lines", () => {
    expect(parseLine("   ")).toBeNull();
  });

  it("skips irrelevant entry types (queue-operation, attachment, mode, last-prompt)", () => {
    const results = fixtureLines.map(parseLine);
    const skippedCount = results.filter((r) => r === null).length;
    // fixture has 4 irrelevant entries: attachment, queue-operation, mode, last-prompt
    expect(skippedCount).toBe(4);
  });

  it("marks the human message entry with isHumanMessage: true", () => {
    const parsed = fixtureLines.map(parseLine).filter((p) => p !== null);
    const humanMessages = parsed.filter((p) => p!.isHumanMessage);
    expect(humanMessages).toHaveLength(2);
    expect(humanMessages.map((p) => p!.uuid)).toEqual(["u1", "u3"]);
  });

  it("does not mark tool_result 'user' entries as human messages", () => {
    const parsed = fixtureLines.map(parseLine).filter((p) => p !== null);
    const toolResultEntry = parsed.find((p) => p!.uuid === "u2");
    expect(toolResultEntry?.isHumanMessage).toBe(false);
  });

  it("extracts model and token usage from assistant entries", () => {
    const parsed = fixtureLines.map(parseLine).filter((p) => p !== null);
    const assistantEntry = parsed.find((p) => p!.uuid === "a2");
    expect(assistantEntry).toMatchObject({
      type: "assistant",
      model: "claude-sonnet-5",
      usage: {
        inputTokens: 1000,
        outputTokens: 200,
        cacheCreationInputTokens: 500,
        cacheReadInputTokens: 0,
      },
    });
  });

  it("does not attach usage to non-assistant entries", () => {
    const parsed = fixtureLines.map(parseLine).filter((p) => p !== null);
    const userEntry = parsed.find((p) => p!.uuid === "u1");
    expect(userEntry?.usage).toBeUndefined();
    expect(userEntry?.model).toBeUndefined();
  });
});

File: packages/intellij-plugin/engine/src/test/kotlin/dev/bredson/tokentracker/engine/claudecode/ParseLineTest.kt
package dev.bredson.tokentracker.engine.claudecode

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull

class ParseLineTest {
    @Test
    fun `returns null for non-JSON lines`() {
        assertNull(parseLine("not json"))
    }

    @Test
    fun `returns null for blank lines`() {
        assertNull(parseLine("   "))
    }

    @Test
    fun `skips irrelevant entry types (attachment, queue-operation, mode, last-prompt)`() {
        val skipped = Fixtures.sampleSessionLines.count { parseLine(it) == null }
        assertEquals(4, skipped)
    }

    @Test
    fun `marks the human message entries with isHumanMessage`() {
        val human = Fixtures.sampleSessionParsed.filter { it.isHumanMessage }
        assertEquals(listOf("u1", "u3"), human.map { it.uuid })
    }

    @Test
    fun `does not mark tool_result user entries as human messages`() {
        val toolResult = Fixtures.sampleSessionParsed.first { it.uuid == "u2" }
        assertFalse(toolResult.isHumanMessage)
    }

    @Test
    fun `extracts model and token usage from assistant entries`() {
        val entry = Fixtures.sampleSessionParsed.first { it.uuid == "a2" }
        assertEquals(EntryType.ASSISTANT, entry.type)
        assertEquals("claude-sonnet-5", entry.model)
        assertEquals(
            ParsedUsage(inputTokens = 1000, outputTokens = 200, cacheCreationInputTokens = 500, cacheReadInputTokens = 0),
            entry.usage,
        )
    }

    @Test
    fun `does not attach usage or model to non-assistant entries`() {
        val user = Fixtures.sampleSessionParsed.first { it.uuid == "u1" }
        assertNull(user.usage)
        assertNull(user.model)
    }

    @Test
    fun `returns null when a grouping field is missing`() {
        assertNull(parseLine("""{"type":"user","sessionId":"s","uuid":"u","cwd":"/p"}"""))
        assertNull(parseLine("""{"type":"user","sessionId":"s","timestamp":"t","cwd":"/p"}"""))
    }

    @Test
    fun `ignores assistant usage when any token field is missing`() {
        val entry = parseLine(
            """{"type":"assistant","sessionId":"s","uuid":"a","timestamp":"t","cwd":"/p",
               "message":{"model":"claude-sonnet-5","usage":{"input_tokens":1,"output_tokens":2}}}""".trimIndent(),
        )
        assertEquals(EntryType.ASSISTANT, entry?.type)
        assertNull(entry?.usage)
        assertNull(entry?.model)
    }
}
